### DISCUSSION TITLE
Logic behind automatic control

### DISCUSSION THREAD

**Discussion by MUN0X** [revk/xESP32-Faikout#192] [Q&A]
First of all, I really like the idea behind this device. 
Especially the auto-mode and the fact that you can couple an external BLE sensor directly to  faikin and have faiking calculate a dynamic setpoint for the daikin unit.

For the last few days I have been racking my brain on how I should set this up correctly but the automode seems to have I mind of his own. Can someone help me understand how this works exactly?
I was under the assumption that you would set a setpoint (autot) and a range (autor) and faiking would then just run the AC (in heating) until the max value is reached. Then it would switch the unit off by lowering the calculated setpoint until the temperature drops below the min value again.
This is exactly what I'm after as this would give long and steady "burns" in heating mode which improves the COP (dramatically)!

In practice this does not appear to be the case and the unit just switches on/off rather randomly around the desired temperature.
If i look at the info via mqtt, the automation never goes "beyond".

**Is there a way to get the behaviour that I described above?**

I figured out how the setpoint value for the AC is calculated and set tpredictt = 0 to disable the prediction stuff.
I played around with the offsets and so on but I cant figure it out.
Edit: I just noticed that the unit started to oscillate around the minimum temperature. Switching heating on when bletemp < min and switching off when bletemp > min. With the BLE sensor reporting with a  resolution of 0.01°C this means that this occurs every 30s or so...

_(On a sidenote, I also don't understand why push10 is a positive value.)_

**Comment by revk** [comment]
OK, the range is what you want, so it heats when below min, not heating until it hits max. When cooling it cools when above max, not cooling until it hits min. It should not go "beyond" ideally. That probably is some of the confusion. The idea is you are comfortable between min and max. The wider you make that the less heating or cooling needed.

`push10` is added to min when heating, and subtracted from `max when cooling.

**Reply by MUN0X** [reply]
The range is what i want but I would like to use the entire range to limit the # of times the outdoor unit switches on/of. According to my own testing this almost doubles the efficiency of my multi-spilt system.

I thought that your code contains a basic PD controller to keep the temperature within the desired range. I was under the presumption that by disabling the differentiating part (tpredictt =0) I would get the desired behaviour.

 I've just realized that the approaching/.../beyond part of your code is actually represents the integrating part of an PID controller. 
Would it be possible to introduce an additional setting, eg. "simplemode", that would allow us to skip this advanced logic?

**Reply by revk** [reply]
Well sort of PD, but also having to second guess what the Daikin will do, hence the deliberate overshoot on the target temperature, which complicated matters. Disabling the predict does make it simpler, but it is still aiming for anywhere within min/max, so heat to min not heat to max. The approaching/beyond also allows for deciding it is way out of range and needs to reverse mode, of increase fan speed. So not just for PID.

The change you want is more fundamental - heat to max, wait to min, heat to max, etc. Like hysteresis on a thermostat, which means it also needs an extra internal state. Obviously I could do this. In my case I wanted to ensure a more stable temperature, and indeed the prediction stuff means it can be very stable within a fraction of a degree. The typical hysteresis logic would oscillate min to max, which I would rather not have.

But I can see you might want this if more worried about efficiency. So it is possible. I'll have to look in to it. It would probably be sensible to maintain the prediction to minimised overshoot of max, or min as well.

**Reply by revk** [reply]
OK, I have added an experimental setting `thermostat`. If you set that it should do what you want. Please let me know if that works. New code release in a couple of minutes.

**Reply by MUN0X** [reply]
Thank you for the fast response and changes! I really appreciate it!

The `thermostat` option definitely improves the behaviour. The unit now heats to the max value and switches off.
Weirdly it switches back on before the minimum value is reached. 
Specifically with MIN/SP/MAX of 18/18.5/19, it heated to 19,05, switched off and switched back on at 18,7°C.
Also, when I enable the unit while the BLE temperature is within the min/max range the unit switches on. Not an issue per se, but it might be more logical for the thermostat to wait until the temperature drops below the minumum

**Reply by revk** [reply]
It should still be applying the prediction stuff if set, and the push stuff if set. Unsetting those may be more what you want.

Initialising the state for thermostat mode when turning on may be possible. I'll take a look.

**Reply by revk** [reply]
OK new issue in a few minutes should start in a not heat/cool state if in the range.

**Reply by MUN0X** [reply]
I did set `tpredictt`, `push10` and `switch10` to 0 before testing the `thermostat` function. It just appears to be aiming for the max now. stopping the unit above max, starting it once it drops below max. Pretty much the same as with the minimum value before.
 
Maybe I misconfigured one of the parameters during testing before? (although i checked and reverted back to defaults listed in faikin.c) Is it possible to revert the board back to defaults?

**Reply by revk** [reply]
There is a factory command...

`command/Faikin/factory XXXXXXXXXXXXFaikin` where the X are the MAC address.

**Reply by MUN0X** [reply]
Sorry I missed that command. 
Factory reset did not appear  to have changed much. Unit still heats to max to then starts/stops every time it crosses the maximum.
Still starts heating when starting while in range

**Reply by revk** [reply]
Hmmm, I'll have to look again, that does seem odd.

**Reply by MUN0X** [reply]
I just had a look at the new code.
I'm not a programmer but I think the issue lies with the state of `daiking.thermostat` Once the current temperature goes above `max`, the `daiking.thermostat` flag starts alternating. This flag is also switched when the min value is crossed.

Would the following fix it?
Inserting the following below line 3027:
`if (thermostat && !daikin.thermostat) daikin.thermostat = 1;`
Replacing the if statement on line 3030 with:
`if (thermostat && (daikin.thermostat &&((hot &&  current > max) || (!hot && current < min))))`

**Reply by revk** [reply]
I'll take another look.

**Reply by revk** [reply]
Yeh, not quite, I think I have it now - new code in a few minutes.

**Reply by MUN0X** [reply]
New code works great! 🥇 
Exactly what I wanted and indeed much simpler then what I suggested.

**Reply by MUN0X** [reply]
Thermostat without prediction offers maximum efficiency. 
Enabling prediction will dampen the fluctuations somewhat which could be useful to deal with milder weather. 
The original is still there to offer maximum comfort at the expense of some efficiency (especially with multi-splits).

**Comment by MUN0X** [comment]
Sorry to come back to this but in the meantime I installed Faikins in all of my units.
I've been playing around with the various options for the automatic controls.

Just to confirm? Whats the best way to disable the prediction and advanced automation to just run the basic thermostat functionality?

I see the following lines in the code:
` 2927        if (tpredicts && !isnan (measured_temp))` ==> tpredicts = 0 to disable prediction?
` 3041        if (!isnan (measured_temp) && !isnan (min) && !isnan (max) && tsample)` ==> tsample = 0 to disable advanced logic?

Is that logic correct or am i overlooking something?

**Reply by revk** [reply]
There is a thermostat setting. And the prediction stuff all has settings that can be set to zeros.

Maybe I need to write up more on it.

**Reply by revk** [reply]
OK each of the settings has a description if you look in the Advanced tab in settings on the web control.

Yes, setting some of these to zero stops the prediction logic.
